src/VisualStudio/Core/Def/ProjectSystem/CPS/IWorkspaceProjectContextFactory.cs

```diff
@@ -42,5 +37,78 @@ internal interface IWorkspaceProjectContextFactory
             string? binOutputPath,
             string? assemblyName,
             CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Creates and initializes a new project and returns a <see
+        /// cref="IWorkspaceProjectContext"/> to lazily initialize the properties and items for the
+        /// project.  This method guarantees that either the project is added (and the returned task
+        /// completes) or cancellation is observed and no project is added.
+        /// </summary>
+        /// <param name="data">Providers access to msbuild evaluation data for the project.</param>
+        /// <param name="hierarchy">The IVsHierarchy for the project; this is used to track linked files across multiple projects when determining contexts.</param>
+        Task<IWorkspaceProjectContext> CreateProjectContextAsync(EvaluationData data, object? hierarchy, CancellationToken cancellationToken);
```

Comment:
Why not declare `hierarchy` as `IVsHierarchy` while we're changing these APIs? It seems like that's the only valid type for this parameter.

Reply:
The project system stores it as `object? HostObject` on `UnconfiguredProjectServices`. I assume it's just pass-through so that there is no dependency on VS Shell. I'd keep it that way for now.

In future I think it would be best if we removed it altogether. Ideally, Roslyn would be able to find IVsHierarchy by the project id from Shell API.